### DISCUSSION TITLE
Is it difficult to implement an "auto" button?

### DISCUSSION THREAD

**Discussion by cdhigh** [OpenHantek/OpenHantek6022#195] [Ideas]
**Is your feature request related to a problem? Please describe.**
Some oscilloscopes have an "auto" button, after pressing it, the oscilloscope will automatically set the Horizontal and vertical parameters(v/div and s/div), ensure that several complete waveforms are displayed on the screen, and waveforms occupy the vertical area as much as possible.
This function is very convenient, whether for novices or experienced people.

**Describe the solution you'd like**
I don’t know how difficult it is to implement this function, and what is the workload.
If it is difficult or the workload is too much, forget it.

**Comment by Ho-Ro** [comment]
I looked into this idea about a year ago and couldn't find an easy and reliable way to implement it, but that doesn't mean it's impossible - maybe I'll get a flash of inspiration when I'm lying on the beach on holiday.

**Comment by cdhigh** [comment]
I haven't started reading source code, I don’t know the specific implementation of the code, so what I say below may be nonsense.
But I was also a professional programmer (many years ago).
I think it's similar to the principle of an automatic transmission multimeter.
First step is auto set the v/Div.
Starting from a 20mv/Div, fetch a package of data from DSO, if the waveform peak value exceeds the maximum value of the window grid, increase the div value to fetch data again, until a peak value is below the limit of window.

Second step is Calculation frequency, Then set the time base to show several complete waveforms.

If the second step is difficult, then the first step can be implemented as a feature "AUTO ZOOM".
For example, double-click the screen to automatically zoom waveform.

**Comment by Nikita-Presnov** [comment]
May be you need in it?
![ishue](https://user-images.githubusercontent.com/54064458/123177493-0c324e00-d48e-11eb-93d9-5f0e4d14fdf5.gif)

**Reply by cdhigh** [reply]
What you mean with this picture?

**Reply by Nikita-Presnov** [reply]
I use the auto on the oscilloscope only to see the signal itself, here I can do it with another function.
By the way, somehow this auto mode generally interfered with me, I stupidly did not look at the value, although I saw the correct signal that was given by MSU. For a long time I could not understand why it did not work, and the leg was simply grounded ...

**Comment by Ho-Ro** [comment]
@cdhigh > This function is very convenient, whether for novices or experienced people.

To be honest, I don't like this "Auto" mode. It may be convenient, but like any automatism, it tempts you to press the button instead of thinking for a moment beforehand. When I measure something, I should have an expectation of what I will see - will it be 10 mV or 500 V - in the latter case, have an HV divider ready **beforehand**. Experienced people do this anyway, and newcomers learn the correct procedure this way. Since the Hantek can only measure +- 5 V input range without clipping anyway, you almost always need a 1:10 probe, which you have to select manually and also set on the unit.

**Reply by cdhigh** [reply]
Disagreed.
This function is an auxiliary function, not a substitute for thinking.
Just like automatic multimeter vs manual multimeter.
Some people don’t like automatic multimeters, but manual multimeters are becoming less and less.
This shows some trends.

**Reply by Ho-Ro** [reply]
Feel free to show a proof-of concept. I develop OpenHantek just for fun and for my needs - and as I do not need *auto* it has very low priority.

**Reply by cdhigh** [reply]
Yes, you are right, we all do it unpaid because we are happy with it.
